nordic: ieee802154: fixes to support 802.15.4 on nRF54H20

piotrkoziar opened the pull request:
The fixes suggested in this PR allow building and running ieee802154-based applications for nRF54H20.

piotrkoziar committed:
dts: nrf54h20: add grtc channel 15 to the pool.

Adds channel 15 to the pool of grtc channels available
for allocation (i.e. with 'z_nrf_grtc_timer_chan_alloc')
on nRF54H20.

The change is motivated by lack of available channels
for the nrf_802154_timestamper when building for nRF54H20.

Signed-off-by: Piotr Koziar <piotr.koziar@nordicsemi.no>

piotrkoziar committed:
modules: hal_nordic: implement hfclk start/stop for nRF54H20.

Clock_control is currently not supported on nRF54H20.
This commit adds new way of handling the hfclk targeted for nRF54H20.

This solution shall be replaced once the clock_control is supported
for nRF54H20.

Signed-off-by: Piotr Koziar <piotr.koziar@nordicsemi.no>

piotrkoziar committed:
ipc: fix return code of icbmsg backend send operation.

This commit fixes the issue where a serialization
error was reported after properly sending a data with 'icbmsg' backend.

The icbmsg send function's return code is set to
the sent data's len as in other backends.
The related docs were fixed and updated.

Signed-off-by: Piotr Koziar <piotr.koziar@nordicsemi.no>

piotrkoziar committed:
modules: hal_nordic: turn off temperature update by default for nRF54H20.

Adds an appropriate condition to the Kconfig as
the temperature driver is not supported for nRF54H20 devices yet.

Without this change, a build with NRF 802.15.4 libraries produces
Kconfig error.

Signed-off-by: Piotr Koziar <piotr.koziar@nordicsemi.no>

piotrkoziar commented:
FYI @ankuns

hubertmis commented on modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_clock_zephyr.c:
```diff
@@ -63,53 +74,27 @@ void nrf_802154_clock_hfclk_stop(void)
-
-	__ASSERT_NO_MSG(mgr != NULL);
+	/* Use register directly, there is no support for that task in nrf_lrcconf_task_trigger */
+	NRF_LRCCONF_RADIO_PD->EVENTS_HFXOSTARTED = 0x0;
+	NRF_LRCCONF_RADIO_PD->TASKS_REQHFXO = 0x1;
```
This will misbehave if there are other HFXO users in this CPU. Is it intentional limitation now? If it is, should it be documented somewhere?

piotrkoziar replied:
Done

piotrkoziar committed:
drivers: ieee802154_nrf5: Use BLE.ADDR instead of FICR to create EUI64.

Use BLE.ADDR to create unique (to some extent) EUI64
on nRF54H20 in some cases inside the IEEE 802.15.4 driver.

The amount of EUI64-s available in such a way is very limited
(~16 million).
However, currently there does not seem to be another feasible way to
get device identifiers on nRF54H20 (such are kept in SICR, to which the
radio core has no access).

Signed-off-by: Piotr Koziar <piotr.koziar@nordicsemi.no>

hubertmis commented on modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_clock_zephyr.c:
```diff
@@ -63,53 +74,27 @@ void nrf_802154_clock_hfclk_stop(void)
-
-	ret = onoff_request(mgr, &lfclk_cli);
-	__ASSERT_NO_MSG(ret >= 0);
+	hfclk_is_running = true;
```
I guess you should wait before you state that the clock is ready. I believe the LRCCONF's event is currently unreliable, but you could use something like `k_sleep()` instead. 1 or 2 ms should be enough in most cases.

piotrkoziar replied:
I added a timer instead of `k_sleep()`

hubertmis commented on modules/hal_nordic/nrf_802154/serialization/platform/nrf_802154_spinel_backend_ipc.c:
```diff
@@ -128,14 +128,13 @@ static void spinel_packet_send_thread_fn(void *arg1, void *arg2, void *arg3)
 		rd_idx = get_rb_idx_plus_1(rd_idx);
 
-		if (ret != expected_ret) {
+		if (ret < 0) {
```
This looks like a workaround for a problem in the icbmsg backend. According to the `ipc_service_send()` function description, it must return the number of sent bytes. If icbmsg backend does not comply with the `ipc_service()` protocol, it shall be fixed. This file should not contain a workaround.

With this workaround, you could miss a case when the backend sends only a part of the message returning a positive number lower than the expected one.

piotrkoziar replied:
I reverted the condition back to its original form and made the icbmsg return sent bytes instead of blocks.

Additionally, I made a verification inside `send_control_message` where verifying sent bytes makes more sense in icbmsg because those 3 bytes there are what we actually send.

ankuns commented on modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_clock_zephyr.c:
```diff
@@ -63,53 +74,27 @@ void nrf_802154_clock_hfclk_stop(void)
-	struct onoff_manager *mgr =
-		z_nrf_clock_control_get_onoff(CLOCK_CONTROL_NRF_SUBSYS_LF);
+	/* Use register directly, there is no support for that task in nrf_lrcconf_task_trigger */
+	NRF_LRCCONF_RADIO_PD->TASKS_STOPREQHFXO=0x1;
```
To satisfy compliance checks.
```suggestion
	NRF_LRCCONF_RADIO_PD->TASKS_STOPREQHFXO = 0x1;
```

hubertmis commented on modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_clock_zephyr.c:
```diff
@@ -63,53 +75,34 @@ void nrf_802154_clock_hfclk_stop(void)
-	ret = onoff_cancel_or_release(mgr, &lfclk_cli);
-	__ASSERT_NO_MSG(ret >= 0);
-	lfclk_is_running = false;
+	k_timer_start(&hfclk_started_timer, K_USEC(1000), K_NO_WAIT);
```
I would replace the magic number with a define, something like `#define MAX_HFXO_RAMP_UP_TIME_US 1000`. I think you can read this value from BICR instead of hardcoding here, but initially, it is ok to keep it as it is.

piotrkoziar replied:
Done

hubertmis commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -126,7 +126,11 @@ static void nrf5_get_eui64(uint8_t *mac)
-#if defined(CONFIG_TRUSTED_EXECUTION_NONSECURE) && defined(NRF_FICR_S)
+#if defined(NRF54H_SERIES)
+	/* Can't access SICR with device id on a radio core. Use BLE.ADDR. */
+	deviceid[0] = NRF_FICR->BLE.ADDR[0];
+	deviceid[1] = NRF_FICR->BLE.ADDR[1];
```
nit: how about using Zephyr's `hwinfo_get_device_id()` instead of these ifdefs?
Of course, can be addressed later. It's just an idea

ankuns replied:
Thanks for this hint. Current implementation of `hwinfo_get_device_id` for nrf54h20 is non-functional. see https://github.com/zephyrproject-rtos/zephyr/blob/8c8e2f6243f147953ce3fadb999c4e2a5faf4f5e/drivers/hwinfo/hwinfo_nrf.c#L25 That's the reason the workaround is ever needed.

piotrkoziar replied:
+ We are not sure if we want to add a dependency on the hwinfo module.

hubertmis replied:
> +We are not sure if we want to add a dependency on the hwinfo module.

I don't see a problem with this dependency in the Zephyr-specific code. It's easier to maintain than dependencies to multiple SoCs.

But again, it's just a hint. If you estimate it's easier for you to maintain SoC-specific defines than to depend on a Zephyr module, it's ok for me.

ankuns commented on subsys/ipc/ipc_service/backends/ipc_icbmsg.c:
```diff
@@ -541,7 +541,7 @@ static int send_release(struct backend_data *dev_data, const uint8_t *buffer,
  *				but it cannot change number of required blocks.
  *
- * @return			O or negative error code.
+ * @return			number of sent blocks or negative error code.
```
I would stick to "non-negative value in case of success or negative error code".
In fact this function returns the result of underlying `send_control_message` which returns the result of underlying `icmsg_send` which returns as you fixed in `icmsg.h` file "Number of sent bytes." where the sent bytes refer rather to the size of control message.

hubertmis replied:
The ipc_service backends must follow the ipc_service API. ipc_service currently requires to return number of sent bytes. If we want to change it, we would need to change ipc_service API. Not only the ipc_icbmsg backend.

ankuns replied:
This is now done within the `send` function, after Piotr's fix. This is the core of the fix. So `send` now follows the API.

`send_block` is internal to the ipc_icbmsg implementation. It can return anything and it would really nice if description of what this function returns met the code.

hubertmis replied:
Right, my mistake.

ankuns commented on subsys/ipc/ipc_service/backends/ipc_icbmsg.c:
```diff
@@ -656,7 +656,7 @@ static int match_bound_msg(struct backend_data *dev_data, size_t rx_block_index,
  * @param[in] ept	Endpoint to use.
  *
- * @return		O or negative error code.
+ * @return		number of sent blocks or negative error code.
```
similar to https://github.com/zephyrproject-rtos/zephyr/pull/73156/files#r1613043527

carlescufi commented:
@piotrkoziar please fix the compliance failures

piotrkoziar commented:
> @piotrkoziar please fix the compliance failures

Done. All tests are green.

PerMac reviewed (approved):
Fixes regression seen in the main when building samples/net/wpanusb and samples/net/wpan_serial on nrf54h20 (e.g. in this PR https://github.com/zephyrproject-rtos/zephyr/actions/runs/9415456827/job/25937895143?pr=73790)
